__init__.py

```diff
@@ -157,18 +175,65 @@ def _account_metadata_pattern(self, tree, metadata_key, pattern, internal, mwr,
         portfolio_data = self._portfolio_data(selected_accounts, internal, mwr, twr)
         return title, portfolio_data
 
+
+    def _process_dividends(self,account,currency):
+        parent_name = ":".join(account.name.split(":")[:-1])
+        result = self.ledger.query_shell.execute_query("select SUM(CONVERT(COST(position),'"+self.operating_currency+"')) AS dividends from HAS_ACCOUNT('Dividend') and HAS_ACCOUNT('"+parent_name+"') where LEAF(account) = '"+currency+"'")
+        dividends = ZERO
+        if len(result[2])>0:
+            for row_cost in result[2]:
+                if len(row_cost.dividends.get_positions())==1: 
+                    dividends+=round(abs(row_cost.dividends.get_positions()[0].units.number),2)
+        return dividends
+
     def _process_node(self, node):
         row = {}
+
         row["account"] = node.name
         row['children'] = []
+        row["last-date"] = None
+        row['PnL'] = ZERO
+        row['Dividends'] = ZERO
         date=self.ledger.end_date
         balance = cost_or_value(node.balance, "at_value", g.ledger.price_map, date=date)
         cost = cost_or_value(node.balance, "at_cost", g.ledger.price_map, date=date)
+        
+        #### ADD Units to the report
+        units = cost_or_value(node.balance, "units", g.ledger.price_map, date=date)
+
+        
+        ### Get row currency 
+        row_currency = None
+        if len(list(units.values())) > 0:
+            row["units"] = list(units.values())[0]
+            row_currency = list(units.keys())[0]
+        #### END of UNITS
+        if row_currency != None and row_currency not in self.ledger.options["operating_currency"]:
+            row['Dividends'] = self._process_dividends(node,row_currency)
+
         if self.operating_currency in balance and self.operating_currency in cost:
             balance_dec = round(balance[self.operating_currency], 2)
             cost_dec = round(cost[self.operating_currency], 2)
             row["balance"] = balance_dec
             row["cost"] = cost_dec
+        
+        #### ADD other Currencies
+        elif self.other_currencies[0] in balance and self.other_currencies[0] in cost:
```

Comment:
self.other_currencies could be an empty set, you can't always access element 0 from it